tokio-trace-serde/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+where
+    S: SerializeStruct,
+{
+    fn record_bool(&mut self, field: &Field, value: bool) {
+        // If previous fields serialized successfully, continue serializing,
+        // otherwise, short-circuit and do nothing.
+        if self.state.is_ok() {
+            self.state = self.serializer.serialize_field(field.name(), &value)
+        }
+    }
+
+    fn record_debug(&mut self, field: &Field, value: &fmt::Debug) {
+        if self.state.is_ok() {
+            self.state = self
+                .serializer
+                .serialize_field(field.name(), &format_args!("{:?}", value))
+        }
+    }
+
+    fn record_u64(&mut self, field: &Field, value: u64) {
+        if self.state.is_ok() {
+            self.state = self.serializer.serialize_field(field.name(), &value)
+        }
+    }
+
+    fn record_i64(&mut self, field: &Field, value: i64) {
+        if self.state.is_ok() {
+            self.state = self.serializer.serialize_field(field.name(), &value)
+        }
+    }
+
+    fn record_str(&mut self, field: &Field, value: &str) {
+        if self.state.is_ok() {
+            self.state = self.serializer.serialize_field(field.name(), &value)
+        }
+    }
+}
+
+impl<S: SerializeStruct> SerdeStructVisitor<S> {
+    /// Completes serializing the visited object, returning `Ok(())` if all
+    /// fields were serialized correctly, or `Error(S::Error)` if a field could
+    /// not be serialized.
+    fn finish(self) -> Result<S::Ok, S::Error> {
+        self.state?;
+        self.serializer.end()
+    }
+}
+
+/// `AsSerde` is a trait that provides the `as_serde` function to types in
+/// `tokio-trace` to allow users to serialize their values.
+pub trait AsSerde<'a> {
+    type Serializable: serde::Serialize + self::sealed::Sealed + 'a;
+
+    /// `as_serde` borrows a `tokio-trace` value and returns the serialized value.
+    fn as_serde(&'a self) -> Self::Serializable;
+}
+
+impl<'a> AsSerde<'a> for tokio_trace_core::Metadata<'a> {
+    type Serializable = SerializeMetadata<'a>;
+
+    fn as_serde(&'a self) -> Self::Serializable {
+        SerializeMetadata(self)
+    }
+}
+
+impl<'a> AsSerde<'a> for tokio_trace_core::Event<'a> {
+    type Serializable = SerializeEvent<'a>;
+
+    fn as_serde(&'a self) -> Self::Serializable {
+        SerializeEvent(self)
+    }
+}
+
+impl<'a> AsSerde<'a> for tokio_trace_core::span::Attributes<'a> {
+    type Serializable = SerializeAttributes<'a>;
+
+    fn as_serde(&'a self) -> Self::Serializable {
+        SerializeAttributes(self)
+    }
+}
+
+impl<'a> AsSerde<'a> for tokio_trace_core::span::Id {
+    type Serializable = SerializeId<'a>;
+
+    fn as_serde(&'a self) -> Self::Serializable {
+        SerializeId(self)
+    }
+}
+
+impl<'a> AsSerde<'a> for tokio_trace_core::span::Record<'a> {
+    type Serializable = SerializeRecord<'a>;
+
+    fn as_serde(&'a self) -> Self::Serializable {
+        SerializeRecord(self)
+    }
+}
+
+impl<'a> self::sealed::Sealed for SerializeMetadata<'a> {}
```

Comment:
Since the `Sealed` bound should be on `AsSerde` rather than `AsSerde::Serializable`, we should change these implementations of `Sealed` to be for `Event`, `Attributes`, `Id`, `Record`, and `Metadata`.